### ISSUE TITLE
systemd-gpt-auto-generator ignores fstab entries

### ISSUE THREAD

**Issue by pyromaniac2k** [systemd/systemd#168]
From the manpage:
“Note that this generator has no effect on non-GPT systems, on systems where the units are explicitly configured (for example, listed in fstab(5))[...]“

In reality, I have my encrypted home partition in both fstab and crypttab, but gpt-auto-generator tries to unlock it at boot, ignoring the noauto directive. This leads to the entire boot process stopping and waiting indefinitely for a password.

**Comment by arvidjaar** [comment]
Yes, I do not see where systemd-gpt-auto-generator would check /etc/fstab. Which likely means that both systemd-gpt-auto-generator and systemd-fstab-generator race for home.mount and the last one wins.

**Comment by xnox** [comment]
you can set a partition flag on the /home partition instead to plug gpt-auto-generator.

**Comment by poettering** [comment]
@arvidjaar s-g-a-g writes into the "late" generator dir, while s-f-g writes in the "normal" generator dir. See https://wiki.freedesktop.org/www/Software/systemd/Generators/ for details. Effect is that s-f-g generated units will override s-g-a-g generated units.

**Comment by poettering** [comment]
@pyromaniac2k i am a bit confused. on one hand you talk about fstab but on the other you talk about a unlocking the home partition. One is handled by systemd-cryptsetup-generator, the other by systemd-fstab-generator. Which one of the two conflicts with systemd-gpt-auto-generator?

Most likely the issue is simply that the s-c-g and s-f-g generated unit files override the s-g-a-g unit files, but the dependencies generated via symlinks are merged.

**Comment by pyromaniac2k** [comment]
@poettering basically I have an encrypted home partition that gets unlocked by pam_mount, so it should not be touched during boot. s-c-g and s-f-g correctly ignore the partition as long as I replace the s-g-a-g binary with /bin/true. When I leave the s-g-a-g binary in place, something tries to unlock the partition during boot, and that something waits for a password without any timeout.

So something causes a start job for dev-mapper-home.device, overriding the "noauto" setting in crypttab.

**Comment by keszybz** [comment]
@poettering yes, there's nothing which prevents symlinks created by gpt-auto-generator from being honored. _If_ we implemented the idea of masking in .wants and .requires directories that Debian people wanted, we could modify fstab-generator to mask .wants or .requires symlinks. Alternatively, we could simply modify gpt-auto-generator to loop over fstab and ignore anything declared there. I think this might be a better alternative actually, because it's less complicated and probably more robust.

**Comment by poettering** [comment]
@pyromaniac2k Hmm, but s-g-a-g will not do anything with home partitions unless they have the gpt type uuid set for home partitions. If you set the uuid you basically tell systemd to automount it (unless you turn off the "automount" bit in the gpt entry again). Hence afaiks there's nothing to fix here: just change the gpt type uuid of your partition to generic linux and there's no automatic mounting done.

Basically you asked for automatic mounting, hence you got it... Don't ask for automatic mounting and you won't have it...

**Comment by pyromaniac2k** [comment]
@poettering When you say that this is correct behaviour, you still have a documentation bug, at least fix the manpage then.

**Comment by poettering** [comment]
@pyromaniac2k did that now in c6355b313e653bc33efe0ed7c047028bb5c42786
